Fix snapshot harness source location initialisation

xbauch opened the pull request:
Before we simply took the first instruction with the line number geq to the one
specified. Now we go through all instructions and take the one closest.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

xbauch commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -504,20 +504,42 @@ memory_snapshot_harness_generatort::initialize_entry_via_source(
 
+  size_t best_match_distance = -1;
+  irep_idt best_match_function_name;
+  std::tuple<goto_programt::const_targett, irep_idt, size_t> best_match;
```
not used.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 72e9499d519128252b2584b7ed4070cd455db43d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/120473357

danpoe commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -504,20 +504,42 @@ memory_snapshot_harness_generatort::initialize_entry_via_source(
+      if(best_match_distance > candidate_instruction.second)
+      {
+        best_match_function_name = entry.first;
+        best_match_distance = candidate_instruction.second;
```
I think we can record `candidate_instruction.first` here as well, then we don't need to look it up again in the code block below.

xbauch replied:
I wrapped those variables in a struct and added the instruction as well.

danpoe commented on regression/snapshot-harness/structs_02/main.c:
```diff
@@ -0,0 +1,41 @@
+  foo();
+
+  return 0;
+}
```
We need a different, smaller regression test. Maybe one in which the source location given to `--initial-source-location` is an empty line in the C program, and thus does not appear in the corresponding goto program.

xbauch replied:
Added.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8fb3cc5f38b97f5e28dd65756bdb1e1fc8983659).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/120839694

petr-bauch committed:
Regression tests for source line input

One where the line exists (there is a goto instruction with that exact line
number) and one where it does not.

petr-bauch committed:
Fix source-location init for snapshot-harness

Before we simply took the first instruction with the line number geq to the one
specified. Now we go through all instructions and take the one closest.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b222e5712e8c6c874013b71724bdd2aa2d0268ed).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/121220124
